### ISSUE TITLE
`Dry::Validation::Schema::Form` doesn't work with nested data

### ISSUE THREAD

**Issue by kirs** [dryrb/dry-validation#41]
```ruby
require 'bundler/setup'
require 'pry'
require 'dry-validation'
require 'dry/validation/schema/form'

class Schema < Dry::Validation::Schema
  key(:phone_numbers) do |phone_numbers|
    phone_numbers.array? do
      phone_numbers.each(&:str?)
    end
  end
end

schema = Schema.new
# works

class FormSchema < Dry::Validation::Schema::Form
  key(:phone_numbers) do |phone_numbers|
    phone_numbers.array? do
      phone_numbers.each(&:str?)
    end
  end
end

schema = FormSchema.new

# /Users/kir/Projects/project/vendor/bundle/ruby/2.2.0/gems/dry-validation-0.4.0/lib/dry/validation/input_type_compiler.rb:34:in `visit': undefined method `visit_each' for #<Dry::Validation::InputTypeCompiler:0x007fde6d48e768> (NoMethodError)
# 	from /Users/kir/Projects/project/vendor/bundle/ruby/2.2.0/gems/dry-validation-0.4.0/lib/dry/validation/input_type_compiler.rb:47:in `block in visit_and'
# 	from /Users/kir/Projects/project/vendor/bundle/ruby/2.2.0/gems/dry-validation-0.4.0/lib/dry/validation/input_type_compiler.rb:47:in `map'
# 	from /Users/kir/Projects/project/vendor/bundle/ruby/2.2.0/gems/dry-validation-0.4.0/lib/dry/validation/input_type_compiler.rb:47:in `visit_and'
# 	from /Users/kir/Projects/project/vendor/bundle/ruby/2.2.0/gems/dry-validation-0.4.0/lib/dry/validation/input_type_compiler.rb:34:in `visit'
# 	from /Users/kir/Projects/project/vendor/bundle/ruby/2.2.0/gems/dry-validation-0.4.0/lib/dry/validation/input_type_compiler.rb:44:in `block in visit_and'
# 	from /Users/kir/Projects/project/vendor/bundle/ruby/2.2.0/gems/dry-validation-0.4.0/lib/dry/validation/input_type_compiler.rb:44:in `map'
# 	from /Users/kir/Projects/project/vendor/bundle/ruby/2.2.0/gems/dry-validation-0.4.0/lib/dry/validation/input_type_compiler.rb:44:in `visit_and'
# 	from /Users/kir/Projects/project/vendor/bundle/ruby/2.2.0/gems/dry-validation-0.4.0/lib/dry/validation/input_type_compiler.rb:34:in `visit'
# 	from /Users/kir/Projects/project/vendor/bundle/ruby/2.2.0/gems/dry-validation-0.4.0/lib/dry/validation/input_type_compiler.rb:29:in `block in call'
# 	from /Users/kir/Projects/project/vendor/bundle/ruby/2.2.0/gems/dry-validation-0.4.0/lib/dry/validation/input_type_compiler.rb:29:in `map'
# 	from /Users/kir/Projects/project/vendor/bundle/ruby/2.2.0/gems/dry-validation-0.4.0/lib/dry/validation/input_type_compiler.rb:29:in `call'
# 	from /Users/kir/Projects/project/vendor/bundle/ruby/2.2.0/gems/dry-validation-0.4.0/lib/dry/validation/schema/form.rb:11:in `initialize'
# 	from bug.rb:24:in `new'
# 	from bug.rb:24:in `<main>'
```

**Comment by solnic** [comment]
I'll fix it in 0.4.1. Thanks for the report

**Comment by solnic** [comment]
This is fixed in 0.4.1 and needs an upgrade of dry-data to 0.4.2 (enforce by dry-validation gemspec)

**Comment by kirs** [comment]
Thanks!
